skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+
+        locations = list(locations)
+        self.locations = locations
+
+        # Boundaries: has to be after locations assignment because
+        # locations setter reset boundaries.
+        indices = [i[0] for i in sorted(enumerate(locations),
+                                        key=operator.itemgetter(1))]
+        # use the boundaries setter to validate the value
+        self.boundaries = boundaries
+        self.boundaries = [self.boundaries[i] for i in indices]
+
+        # Metadata
+        if metadata is None:
+            metadata = {}
+        self.metadata = metadata
+
+        # add this interval feature to the associated IntervalMetadata
+        self._add()
+
+    def _add(self):
+        """Add the current ``Interval`` to the IntervalMetadata object."""
+        for loc in self.locations:
+            start, end = loc
+            self._interval_metadata._interval_tree.add(start, end, self)
+        self._interval_metadata._intervals.append(self)
+
+    @experimental(as_of='0.5.0-dev')
+    def __eq__(self, other):
+        '''Test if this ``Interval`` object is equal to another.
+
+        The equality is performed by checking if the ``metadata``,
+        ``location`` and ``boundaries`` are equal.
+
+        Parameters
+        ----------
+        other : Interval
+            Interval to test for equality against.
+
+        Returns
+        -------
+        bool
+            Indicates if the two objects are equal.
+        '''
+        return ((self.metadata == other.metadata) and
+                (self.locations == other.locations) and
+                (self.boundaries == other.boundaries))
+
+    @experimental(as_of='0.5.0-dev')
+    def __ne__(self, other):
+        '''Test if this ``Interval`` object is not equal to another.
+
+        Parameters
+        ----------
+        other : Interval
+            Interval to test for inequality against.
+
+        Returns
+        -------
+        bool
+            Indicates if the two objects are not equal.
+        '''
+        return not self.__eq__(other)
+
+    @experimental(as_of='0.5.0-dev')
+    def __repr__(self):
+        '''Return a string representation of this ``Interval`` object.
+
+        Returns
+        -------
+        str
+            String representation of this ``Interval`` object.
+        '''
+        s = ('{}(interval_metadata=<{!r}>, locations={!r}, '
+             'boundaries={!r}, metadata={!r})')
+        return s.format(self.__class__.__name__, id(self._interval_metadata),
+                        self.locations, self.boundaries, self.metadata)
+
+    @experimental(as_of='0.5.0-dev')
+    def drop(self):
+        '''Drop this ``Interval`` object from the interval metadata it links to.
+
+        See Also
+        --------
+        skbio.metadata.IntervalMetadata.drop
+        '''
+        self._interval_metadata.drop(locations=self.locations,
+                                     metadata=self.metadata)
```

Comment:
The doc of `IntervalMetadata.drop`  says it uses `query` to get the `Interval` objects to delete, to allow flexible deletion. It is not deleting an object exact equal to an interval object.